File: apps/mobile/src/features/trial/TrialChallengeService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

export type TrialDayType = 1 | 2 | 3;

export type TrialTaskType = 'focusSession' | 'beatFocusTime' | 'missions' | 'disciplineReport';

export interface TrialDay {
  day: TrialDayType;
  focusSessionDone: boolean;
  missionsDone: boolean;
  focusMinutes: number;
}

export interface TrialChallengeState {
  startDate: string;
  days: TrialDay[];
  completed: boolean;
}

const STORAGE_KEY = '@lockedin/trial_challenge';
const TRIAL_DURATION_DAYS = 3;

export class TrialChallengeService {
  /**
   * Initialize a new trial challenge
   */
  static async initializeTrial(): Promise<void> {
    const state: TrialChallengeState = {
      startDate: new Date().toISOString(),
      days: [
        { day: 1, focusSessionDone: false, missionsDone: false, focusMinutes: 0 },
        { day: 2, focusSessionDone: false, missionsDone: false, focusMinutes: 0 },
        { day: 3, focusSessionDone: false, missionsDone: false, focusMinutes: 0 },
      ],
      completed: false,
    };
    await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(state));
  }

  /**
   * Check if the trial is currently active
   * @returns true if within 3 days of trial start
   */
  static async isTrialActive(): Promise<boolean> {
    const state = await this.getState();
    if (!state) return false;

    const startDate = new Date(state.startDate);
    const now = new Date();
    const daysDiff = Math.floor(
      (now.getTime() - startDate.getTime()) / (1000 * 60 * 60 * 24)
    );

    return daysDiff < TRIAL_DURATION_DAYS && !state.completed;
  }

  /**
   * Get the current trial day (1, 2, or 3)
   * @returns trial day number, or 0 if trial is not active
   */
  static async getTrialDay(): Promise<TrialDayType | 0> {
    const isActive = await this.isTrialActive();
    if (!isActive) return 0;

    const state = await this.getState();
    if (!state) return 0;

    const startDate = new Date(state.startDate);
    const now = new Date();
    const daysDiff = Math.floor(
      (now.getTime() - startDate.getTime()) / (1000 * 60 * 60 * 24)
    );

    const currentDay = (daysDiff + 1) as TrialDayType;
    return currentDay <= TRIAL_DURATION_DAYS ? currentDay : (0 as any);
  }

  /**
   * Update focus minutes for a specific day
   */
  static async updateFocusMinutes(day: TrialDayType, minutes: number): Promise<void> {
    const state = await this.getState();
    if (!state) return;

    const dayData = state.days.find(d => d.day === day);
    if (dayData) {
      dayData.focusMinutes = Math.max(dayData.focusMinutes, minutes);
      await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(state));
    }
  }

  /**
   * Mark a task as complete for a specific day
   */
  static async completeTrialTask(day: TrialDayType, task: TrialTaskType): Promise<void> {
    const state = await this.getState();
    if (!state) return;

    const dayData = state.days.find(d => d.day === day);
    if (!dayData) return;

    switch (task) {
      case 'focusSession':
        dayData.focusSessionDone = true;
        break;
      case 'beatFocusTime':
        dayData.focusSessionDone = true;
        break;
      case 'missions':
        dayData.missionsDone = true;
        break;
      case 'disciplineReport':
        // All tasks complete on day 3
        if (day === 3) {
          state.completed = true;
        }
        break;
    }

    await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(state));
  }

  /**
   * Get trial progress
   */
  static async getTrialProgress(): Promise<{
    day: number;
    tasksCompleted: number;
    totalTasks: number;
  }> {
    const currentDay = await this.getTrialDay();
    if (currentDay === 0) {
      return { day: 0, tasksCompleted: 0, totalTasks: 0 };
    }

    const state = await this.getState();
    if (!state) {
      return { day: currentDay, tasksCompleted: 0, totalTasks: 0 };
    }

    const dayData = state.days.find(d => d.day === currentDay);
    if (!dayData) {
      return { day: currentDay, tasksCompleted: 0, totalTasks: 0 };
    }

    let tasksCompleted = 0;
    let totalTasks = 0;

    if (currentDay === 1) {
      totalTasks = 1;
      if (dayData.focusSessionDone) tasksCompleted += 1;
    } else if (currentDay === 2) {
      totalTasks = 2;
      if (dayData.focusSessionDone) tasksCompleted += 1;
      if (dayData.missionsDone) tasksCompleted += 1;
    } else if (currentDay === 3) {
      totalTasks = 1; // Discipline Report check
      if (dayData.focusSessionDone) tasksCompleted += 1; // Simplified: treat completion as viewed report
    }

    return { day: currentDay, tasksCompleted, totalTasks };
  }

  /**
   * Get time until trial expires
   * @returns Object with hours and minutes remaining
   */
  static async getTrialTimeRemaining(): Promise<{
    hours: number;
    minutes: number;
    expired: boolean;
  }> {
    const state = await this.getState();
    if (!state) {
      return { hours: 0, minutes: 0, expired: true };
    }

    const startDate = new Date(state.startDate);
    const trialEndDate = new Date(startDate.getTime() + TRIAL_DURATION_DAYS * 24 * 60 * 60 * 1000);
    const now = new Date();

    if (now >= trialEndDate) {
      return { hours: 0, minutes: 0, expired: true };
    }

    const totalMinutesRemaining = Math.floor(
      (trialEndDate.getTime() - now.getTime()) / (1000 * 60)
    );
    const hours = Math.floor(totalMinutesRemaining / 60);
    const minutes = totalMinutesRemaining % 60;

    return { hours, minutes, expired: false };
  }

  /**
   * Reset trial (clear all progress)
   */
  static async resetTrial(): Promise<void> {
    await AsyncStorage.removeItem(STORAGE_KEY);
  }

  /**
   * Get current trial state
   */
  private static async getState(): Promise<TrialChallengeState | null> {
    try {
      const stored = await AsyncStorage.getItem(STORAGE_KEY);
      if (!stored) return null;

      return JSON.parse(stored) as TrialChallengeState;
    } catch (error) {
      console.error('Failed to get trial challenge state:', error);
      return null;
    }
  }
}

File: apps/ios-swift/LockedIn/Features/Trial/TrialChallengeService.swift
import Foundation

/// Swift port of `apps/mobile/src/features/trial/TrialChallengeService.ts`.
///
/// Local-only (no Supabase). Persists a 3-day challenge state at
/// `@lockedin/trial_challenge` in standard `UserDefaults`.
///
/// API mirrors the RN static-class surface 1:1:
/// - `initializeTrial()`
/// - `isTrialActive()` -> Bool
/// - `getTrialDay()` -> 0 | 1 | 2 | 3
/// - `updateFocusMinutes(day:minutes:)`
/// - `completeTrialTask(day:task:)`
/// - `getTrialProgress()` -> (day, tasksCompleted, totalTasks)
/// - `getTrialTimeRemaining()` -> (hours, minutes, expired)
/// - `resetTrial()`
public enum TrialDayType: Int, Codable, Sendable {
    case one = 1
    case two = 2
    case three = 3
}

public enum TrialTaskType: String, Sendable {
    case focusSession
    case beatFocusTime
    case missions
    case disciplineReport
}

public struct TrialDay: Codable, Sendable, Equatable {
    public var day: Int
    public var focusSessionDone: Bool
    public var missionsDone: Bool
    public var focusMinutes: Int

    public init(day: Int, focusSessionDone: Bool = false, missionsDone: Bool = false, focusMinutes: Int = 0) {
        self.day = day
        self.focusSessionDone = focusSessionDone
        self.missionsDone = missionsDone
        self.focusMinutes = focusMinutes
    }
}

public struct TrialChallengeState: Codable, Sendable, Equatable {
    public var startDate: String  // ISO 8601 string — matches RN `new Date().toISOString()`
    public var days: [TrialDay]
    public var completed: Bool

    public init(startDate: String, days: [TrialDay], completed: Bool) {
        self.startDate = startDate
        self.days = days
        self.completed = completed
    }
}

public struct TrialProgress: Sendable, Equatable {
    public let day: Int
    public let tasksCompleted: Int
    public let totalTasks: Int
}

public struct TrialTimeRemaining: Sendable, Equatable {
    public let hours: Int
    public let minutes: Int
    public let expired: Bool
}

public enum TrialChallengeService {
    /// Persistence key — preserve verbatim from RN.
    public static let storageKey = "@lockedin/trial_challenge"
    /// Total trial length.
    public static let trialDurationDays = 3

    // MARK: - ISO 8601 formatter (matches JS `toISOString()`)

    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static func isoString(from date: Date) -> String {
        isoFormatter.string(from: date)
    }

    private static func date(fromISO string: String) -> Date? {
        // ISO8601DateFormatter with fractional seconds is strict — fall back
        // to a forgiving parser if the stored string lacks ms precision.
        if let d = isoFormatter.date(from: string) { return d }
        let fallback = ISO8601DateFormatter()
        fallback.formatOptions = [.withInternetDateTime]
        return fallback.date(from: string)
    }

    // MARK: - Public API

    /// Initialize a new trial challenge state, persisting it immediately.
    public static func initializeTrial() {
        let state = TrialChallengeState(
            startDate: isoString(from: Date()),
            days: [
                TrialDay(day: 1),
                TrialDay(day: 2),
                TrialDay(day: 3)
            ],
            completed: false
        )
        Defaults.setCodable(state, storageKey)
    }

    /// Returns true if within `trialDurationDays` of start and not completed.
    public static func isTrialActive() -> Bool {
        guard let state = getState() else { return false }
        guard let start = date(fromISO: state.startDate) else { return false }

        let daysDiff = Int(floor(Date().timeIntervalSince(start) / (60.0 * 60.0 * 24.0)))
        return daysDiff < trialDurationDays && !state.completed
    }

    /// Current trial day (1, 2, or 3); 0 if not active.
    public static func getTrialDay() -> Int {
        guard isTrialActive() else { return 0 }
        guard let state = getState(), let start = date(fromISO: state.startDate) else { return 0 }

        let daysDiff = Int(floor(Date().timeIntervalSince(start) / (60.0 * 60.0 * 24.0)))
        let currentDay = daysDiff + 1
        return currentDay <= trialDurationDays ? currentDay : 0
    }

    /// Update focus minutes for a specific day (only ever increases).
    public static func updateFocusMinutes(day: TrialDayType, minutes: Int) {
        guard var state = getState() else { return }
        guard let idx = state.days.firstIndex(where: { $0.day == day.rawValue }) else { return }
        state.days[idx].focusMinutes = max(state.days[idx].focusMinutes, minutes)
        Defaults.setCodable(state, storageKey)
    }

    /// Mark a task as complete for a specific day.
    public static func completeTrialTask(day: TrialDayType, task: TrialTaskType) {
        guard var state = getState() else { return }
        guard let idx = state.days.firstIndex(where: { $0.day == day.rawValue }) else { return }

        switch task {
        case .focusSession, .beatFocusTime:
            state.days[idx].focusSessionDone = true
        case .missions:
            state.days[idx].missionsDone = true
        case .disciplineReport:
            if day == .three {
                state.completed = true
            }
        }

        Defaults.setCodable(state, storageKey)
    }

    /// Trial progress summary for the current day.
    public static func getTrialProgress() -> TrialProgress {
        let currentDay = getTrialDay()
        if currentDay == 0 {
            return TrialProgress(day: 0, tasksCompleted: 0, totalTasks: 0)
        }
        guard let state = getState(),
              let dayData = state.days.first(where: { $0.day == currentDay }) else {
            return TrialProgress(day: currentDay, tasksCompleted: 0, totalTasks: 0)
        }

        var tasksCompleted = 0
        var totalTasks = 0
        switch currentDay {
        case 1:
            totalTasks = 1
            if dayData.focusSessionDone { tasksCompleted += 1 }
        case 2:
            totalTasks = 2
            if dayData.focusSessionDone { tasksCompleted += 1 }
            if dayData.missionsDone { tasksCompleted += 1 }
        case 3:
            totalTasks = 1
            if dayData.focusSessionDone { tasksCompleted += 1 }
        default: break
        }
        return TrialProgress(day: currentDay, tasksCompleted: tasksCompleted, totalTasks: totalTasks)
    }

    /// Time remaining until the trial expires.
    public static func getTrialTimeRemaining() -> TrialTimeRemaining {
        guard let state = getState(), let start = date(fromISO: state.startDate) else {
            return TrialTimeRemaining(hours: 0, minutes: 0, expired: true)
        }

        let trialEnd = start.addingTimeInterval(Double(trialDurationDays) * 24.0 * 60.0 * 60.0)
        let now = Date()
        if now >= trialEnd {
            return TrialTimeRemaining(hours: 0, minutes: 0, expired: true)
        }

        let totalMinutesRemaining = Int(floor(trialEnd.timeIntervalSince(now) / 60.0))
        let hours = totalMinutesRemaining / 60
        let minutes = totalMinutesRemaining % 60
        return TrialTimeRemaining(hours: hours, minutes: minutes, expired: false)
    }

    /// Clear all trial progress.
    public static func resetTrial() {
        Defaults.remove(storageKey)
    }

    /// Read-only access to the persisted state (used by the UI layer).
    public static func currentState() -> TrialChallengeState? {
        getState()
    }

    // MARK: - Private

    private static func getState() -> TrialChallengeState? {
        Defaults.codable(TrialChallengeState.self, storageKey)
    }
}
